docs/transpile/custom-transpiler-pass.ipynb

```diff
@@ -0,0 +1,239 @@
+   "metadata": {},
+   "source": [
+    "## Pauli Twirling\n",
+    "\n",
+    "As an example we construct a transpiler pass that adds Pauli twirls. Pauli twirling is an error suppression strategy that randomizes how the qubits experience noisy channels, which we assume to be two-qubit gates below (because they are much more error-prone than single-qubit gates). However, the Pauli twirls do not effect the actual operation of the two-qubit gate, they are chosen such that those applied *before* the two-qubit gate (to the left) are countered by those applied *after* the two-qubit gate (to the right). In this sense, the two-qubit operations are identical but the way they are performed is different. One benefit of Pauli twirling is that it turns coherent errors into stochastic errors, which can be improved by averaging more.\n",
+    "\n",
+    "Transpiler passes act on the DAG (directed acyclic graph **TODO: include link**), and the important method to override is `.run()`, which takes the DAG as the input. Here we will also add the initialization of pairs of Paulis that preserve the operation of each two-qubit gate. This is done with the helped method `build_twirl_set` that goes through each two-qubit Pauli (as obtained from `pauli_basis(2)`) and finding the other Pauli that preserves the operation. \n",
+    "\n",
+    "From the DAG we can collect runs/nodes (**TODO: what really is the difference between these**), which provides the set of every gate with a given name in the DAG. **TODO: what other ways can you search, like `.bfs_successors` and such? Does this belong here or in the DAG doc?**. Here, we wish to replace each node, representing an instruction, with a subcircuit which we build as a mini DAG. The mini DAG has a two-qubit quantum register added to it, and operations are added here with `apply_operation_back`, which places the `Instruction` on the output of the mini DAG (whereas `apply_operation_front` would place it on the input of the mini DAG). The node is then substituted by the mini DAG with `substitute_node_with_dag`, and the process continues over each instance of `CXGate` and `ECRGate` in the DAG (corresponding to the two-qubit basis gates on IBM backends)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qiskit.dagcircuit import DAGCircuit\n",
+    "from qiskit.circuit import Gate, QuantumCircuit, QuantumRegister\n",
+    "from qiskit.circuit.library import CXGate, ECRGate\n",
+    "from qiskit.transpiler.basepasses import TransformationPass\n",
+    "from qiskit.quantum_info import Operator, pauli_basis\n",
+    "\n",
+    "import numpy as np\n",
+    "\n",
+    "from typing import Iterable, Optional"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class PauliTwirl(TransformationPass):\n",
+    "    \"\"\"Add Pauli twirls to two-qubit gates.\"\"\"\n",
```

Comment:
could add some code comments into the code block explaining what's happening in the code itself? I think this would help readers match up what's happening in the code example with the explanation above it